utils/relay-manager.js

```diff
@@ -61,11 +91,119 @@ class RelayManager {
+      { gas, gasPrice: _gasPrice, from: relayerAccount },
     );
 
+    // console.log("_gasLimit", _gasLimit);
+    // console.log("gas sent", gas);
+    // console.log("gas used", tx.receipt.gasUsed);
+    // console.log("ratio", _gasLimit / tx.receipt.gasUsed);
+
     return tx.receipt;
   }
+
+  /* Returns the gas limit to use as gasLimit parameter in execute function
+    + 1856  (isFeatureAuthorisedInVersionManager check)
+    + 0 / 1000 / 4800 based on which contract implements getRequiredSignatures()
+    + 2052 (getSignHash call)
+    + 45144 (checkAndUpdateUniqueness call)
+    + 10000 * number of signatures (validateSignatures call, should best be estimated but this is also close enough)
+    + Function call estimate
+    + 40000 / 30000 refund cost for 1 signatures and >1 signatures respectively
+
+    Ignoring multiplication and comparisson as that is <10 gas per operation
+  */
+  async getGasLimitRefund(_module, _method, _params, _wallet, _signers, _gasPrice, _refundToken) {
+    let requiredSigsGas = 0;
+    const { contractName } = _module.constructor;
+    if (contractName === "ApprovedTransfer" || contractName === "RecoveryManager") {
+      requiredSigsGas = 4800;
+    } else if (contractName === "GuardianManager") {
+      requiredSigsGas = 1000;
+    }
+
+    // Estimate cost of checkAndUpdateUniqueness call
+    let nonceCheckGas = 0;
+    if (_signers.length === 1) {
+      nonceCheckGas = 6200;
+    } else if (_signers.length > 1) {
+      nonceCheckGas = 22200;
+    }
+
+    let gasEstimateFeatureCall = 0;
+    try {
+      gasEstimateFeatureCall = await _module.contract.methods[_method](..._params).estimateGas({ from: this.relayerManager.address });
+      gasEstimateFeatureCall -= 21000;
+    } catch (err) { // eslint-disable-line no-empty
+    } finally {
+      // When we can't estimate the inner feature call correctly, set this to some large number
+      // This only happens for the following tests atm:
+      // approvedTransfer should revert when target contract is an authorised module
+      // simpleUpgrader should not upgrade to 0 module (relayed tx)
+      // nftTransfer should allow safe NFT transfer from wallet1 to wallet2 (relayed)
+      if (gasEstimateFeatureCall <= 0) {
+        gasEstimateFeatureCall = 140000;
+      }
+    }
+
+    let refundGas = 0;
+    const methodData = _module.contract.methods[_method](..._params).encodeABI();
+    const requiredSignatures = await _module.getRequiredSignatures(_wallet.address, methodData);
+
+    if (_gasPrice > 0 && requiredSignatures[1].toNumber() === 1) {
+      if (_signers.length > 1) {
+        refundGas = 30000;
+      } else {
+        refundGas = 40000;
+      }
+
+      // If the refund is with a token, add transfer cost.
+      // We are using a simple ERC20 transfer cost, however this varies by supported token, e.g. ZRX, BAT, REP, DAI, USDC, or USDT
+      if (_refundToken !== ETH_TOKEN) {
+        refundGas += 30000;
```

Comment:
I don't understand where this additional 30k comes from.

Reply:
This was the option for using better relayer-level gasLimit estimate instead of adding the 50,000 buffer. I've commented it out with some description in case we decide to come back to it.